### PR TITLE
Add metrics to attributes

### LINKED ISSUE #64: Spans of font sizes
This is a highly requested feature from a rich text editor - being able to change font size within a line.

I looked into doing this myself but couldn't figure out a good way to get the attrs into layout without adding a font size field to the shaping structs, which seemed to be against the spirit of the design. If there's any advice on what changes would need to be made where to make this happen, I'll gladly do it.

### PR DESCRIPTION
**Pull request by jackpot51**
Rewrite of #235, fixes #64

### CODE REVIEW

File: src/attrs.rs [hunk 1]
```diff
@@ -101,6 +101,32 @@ impl FamilyOwned {
     }
 }
 
+/// Metrics, but implementing Eq and Hash using u32 representation of f32
+//TODO: what are the edge cases of this?
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct CacheMetrics {
+    pub font_size_bits: u32,
```
Comment (alice-i-cecile) [inline src/attrs.rs:108]: I would be very reluctant to expose these fields directly. Getters / setters seem like the way to go here.

File: src/shape.rs [hunk 2]
```diff
@@ -513,15 +518,17 @@ impl ShapeGlyph {
             cache_key_flags: self.cache_key_flags,
         }
     }
+
+    pub fn width(&self, font_size: f32) -> f32 {
```
Comment (alice-i-cecile) [inline src/shape.rs:522]: Doc strings would be helpful here. Units are the main question.

File: src/attrs.rs (hunk 1 above)
Comment (jackpot51) [reply src/attrs.rs:108]: I think just the into/from implementations for Metrics are enough.

**Review by nicoburns — CHANGES_REQUESTED** [review]
Haven't had chance to review the code yet, but there is bug when creating new lines:

Screenshot 2024-06-06 at 23 41 06

(newly created lines end up with the "font size 64" line-height even though there is no text on that line with that height)

**Comment by jackpot51** [comment]
Right now it takes the maximum of either the buffer line height or the glyph line height for the whole run. I could probably only reference the glyph line height but I didn't want to run into empty lines being 0 height.

**Comment by jackpot51** [comment]
@nicoburns should be resolved in the latest commit.

**Review by alice-i-cecile — APPROVED** [review]
This seems reasonable from a code quality perspective, and I can follow the logic :)

**Comment by jackpot51** [comment]
This is working well enough for me that I will merge. Any issues I missed can be reported against master.

**Pull request merged**